pkg/reconciler/gc/v2/gc.go

```diff
@@ -61,51 +64,67 @@ func Collect(
+	stale, nonactive := 0, 0
+	for _, rev := range revs {
+		if isRevisionActive(rev, config) {
+			continue
+		}
+		nonactive++
+		if isRevisionStale(cfg, rev, logger) {
+			stale++
+		}
+
+		if stale > min {
+			logger.Infof("Deleting stale revision %q", rev.ObjectMeta.Name)
+		} else if max != gc.Disabled && nonactive > max {
+			logger.Infof("Maximum(%d) reached. Deleting oldest non-active revision %q",
+				max, rev.ObjectMeta.Name)
+		} else {
 			continue
 		}
-		if isRevisionStale(ctx, rev, config) {
-			err := client.ServingV1().Revisions(rev.Namespace).Delete(rev.Name, &metav1.DeleteOptions{})
-			if err != nil {
-				logger.With(zap.Error(err)).Errorf("Failed to delete stale revision %q", rev.Name)
-				continue
-			}
+
+		if err := client.ServingV1().Revisions(rev.Namespace).Delete(
+			rev.Name, &metav1.DeleteOptions{}); err != nil {
+			logger.With(zap.Error(err)).Errorf("Failed to delete stale revision %q", rev.Name)
+			continue
 		}
 	}
 	return nil
 }
 
-func isRevisionStale(ctx context.Context, rev *v1.Revision, config *v1.Configuration) bool {
+func isRevisionActive(rev *v1.Revision, config *v1.Configuration) bool {
 	if config.Status.LatestReadyRevisionName == rev.Name {
-		return false
+		return true // never delete latest ready, even if config is not active.
 	}
-	cfg := configns.FromContext(ctx).RevisionGC
-	logger := logging.FromContext(ctx)
-	curTime := time.Now()
-	createTime := rev.ObjectMeta.CreationTimestamp
 
-	if createTime.Add(cfg.StaleRevisionCreateDelay).After(curTime) {
-		// Revision was created sooner than staleRevisionCreateDelay. Ignore it.
-		return false
+	if strings.EqualFold(rev.Annotations[serving.RevisionPreservedAnnotationKey], "true") {
+		return true
 	}
+	// Anything that the labeler hasn't explicitly labelled as inactive.
+	// Revisions which do not yet have any annotation are not eligible for deletion.
+	return rev.GetRoutingState() != v1.RoutingStateReserve
+}
 
-	lastActive := revisionLastActiveTime(rev)
+func isRevisionStale(cfg *gc.Config, rev *v1.Revision, logger *zap.SugaredLogger) bool {
+	sinceCreate, sinceActive := cfg.RetainSinceCreateTime, cfg.RetainSinceLastActiveTime
+	if sinceCreate == gc.Disabled && sinceActive == gc.Disabled {
+		return false // Time checks are both disabled. Not stale.
+	}
 
-	// TODO(whaught): this is carried over from v1, but I'm not sure why we can't delete a ready revision
-	// that isn't referenced? Maybe because of labeler failure - can we replace this with 'pending' routing state check?
-	if lastActive.Equal(createTime.Time) {
-		// Revision was never active and it's not ready after staleRevisionCreateDelay.
-		// It usually happens when ksvc was deployed with wrong configuration.
-		return !rev.Status.GetCondition(v1.RevisionConditionReady).IsTrue()
+	curTime := time.Now()
+	createTime := rev.ObjectMeta.CreationTimestamp
+	if sinceCreate != gc.Disabled && createTime.Add(sinceCreate).After(curTime) {
```

Comment:
`time.Since(createTime) > sinceCreate`.
Same below, is easier to parse, I think.